### ISSUE TITLE
Hash collisions of item stacks causing wrong items to be stored

### ISSUE THREAD

**Issue by kurrycat2004** [AE2-UEL/Applied-Energistics-2#501]
**Describe the bug**

`AEItemStackRegistry#getRegisteredStack(ItemStack)` only compares the hash of the input stack, without comparing the actual item stacks, making it vulnerable to hash collisions. This causes new items put into a system to "transform" into the already existing item if it has the same hash.

**To Reproduce**

- make a simple ME system
- put a Protection 1 enchanted book in
- put a Fire Protection 2 enchanted book in
- system now contains 2 Protection 1 books

**Expected behavior**

ME System correctly stores one Protection 1 book and one Fire Protection 2 book

**Additional context**

https://github.com/user-attachments/assets/3aed0a36-1bb4-43d9-858f-045df4b6775b

**Environment**

Tested with only AE2

- Minecraft Version: 1.12.2
- AE2 Version: v0.56.6
- Forge Version: 14.23.5.2847

**Comment by PukPukov** [comment]
Is this an issue in original AE?

**Comment by kurrycat2004** [comment]
from what i could see, no

**Comment by Exaxxion** [comment]
For what it's worth, 0.56.6 has known issues and I was advised months ago not to update to it because of that.

**Comment by ghzdude** [comment]
this is concerning for CEu, since it uses the same kind of hash strategy

**Comment by ghzdude** [comment]
CEu uses [this](https://github.com/GregTechCEu/GregTech/blob/master/src/main/java/gregtech/api/util/ItemStackHashStrategy.java) in OpenCustomHashMaps (from fastutil), i'm not sure if those use both hash and identity equals.

**Comment by IntegerLimit** [comment]
Yes, that should also check identity equals. AE2 seems to use this: https://github.com/AE2-UEL/Applied-Energistics-2/commit/724f3baf4c8542d8ea7518b90321eb9434c1ef7e, a 'Weak Values' Map from Google. Not sure if it is what I thought.

**Comment by IntegerLimit** [comment]
This is quite a large issue, and I'm of the opinion that perhaps https://github.com/AE2-UEL/Applied-Energistics-2/commit/724f3baf4c8542d8ea7518b90321eb9434c1ef7e should just be reverted, as the performance gains seem to be minimal, and fixing this issue would require direct equality comparison of AESharedItemStack or similar, essentially leading to the same code as before.

Alternatively, the map could use some other ItemStack Representing Object as the key (like ItemTagMeta in GregTech and Nomi Labs).

(Yes, this issue still occurs in latest AE2)

**Comment by xJon** [comment]
Any update on this?
